### ISSUE TITLE
Track LICENSE and NOTICE issues for `incubator-openwhisk`

### ISSUE THREAD

**Issue by mrutkows** [apache/incubator-openwhisk-release#93]
#Please use this issue to track any issues surrounding either the [LICENSE.txt](https://github.com/apache/incubator-openwhisk/blob/master/LICENSE.txt) or [NOTICE.txt](https://github.com/apache/incubator-openwhisk/blob/master/NOTICE.txt) file for the associated repository listed in the title.

**Open Issues/PRs**:
- Issue [#122 - Documentation requirements for LogBack (default logger)](https://github.com/apache/incubator-openwhisk-release/issues/122) - (release repo.)
- Issue [#3607 - Change the JDK to use openjdk8-openj9](https://github.com/apache/incubator-openwhisk/issues/3607) 
- Issue [#137 - JDK license issue for the runtime Java](https://github.com/apache/incubator-openwhisk-release/issues/137)

**Closed Issues/Merged PRs**:
- PR [#3550 - Add the licenses of the dependencies into LICENSE](https://github.com/apache/incubator-openwhisk/pull/3550)
- PR [#3522 - Update notice file](https://github.com/apache/incubator-openwhisk/pull/3522)
    - [Review NOTICE file for main openwhisk repo. #75](https://github.com/apache/incubator-openwhisk-release/issues/75)
- PR [#3565 - Add license headers to files under ansible folder](https://github.com/apache/incubator-openwhisk/pull/3565)
    - Issue [#3264 - A list of source files miss Apache license headers](https://github.com/apache/incubator-openwhisk/issues/3264) 
- PR [#3566 - Add license headers to 67 files](https://github.com/apache/incubator-openwhisk/pull/3566)
    - Issue [#3555 - Do test files need license headers?](https://github.com/apache/incubator-openwhisk/issues/3555)
- PR [#3596 - Add short license header to test data files](https://github.com/apache/incubator-openwhisk/pull/3596#pullrequestreview-116928826)
- PR [#3601 - Change the LICENSE file into the correct format defined in openwhisk release](https://github.com/apache/incubator-openwhisk/pull/3601)
- PR [#3602 - Update build to use latest scancode .cfg for .md and .sh](https://github.com/apache/incubator-openwhisk/pull/3602)
- PR [#3624 - Replace WALA Pair with clean new Pair.](https://github.com/apache/incubator-openwhisk/pull/3624)
    - PR [#3620 - Remove Eclipse license header and IBM copyright from a test file](https://github.com/apache/incubator-openwhisk/pull/3620) (closed)
    - Issue [Replace or remove Pair.java from main openwhisk repo. #142](https://github.com/apache/incubator-openwhisk-release/issues/142)
- PR [#3626 - Add ASF license headers to all Scala .conf files.](https://github.com/apache/incubator-openwhisk/pull/3626)
- PR [#3659 - Assure all Dockerfiles have ASF header.](https://github.com/apache/incubator-openwhisk/pull/3659)
- PR [#3666 - Use Minified ASF License header in Ansible YAML files.](https://github.com/apache/incubator-openwhisk/pull/3666)
    - Issue [ #3660 - Python (*.py) and YAML (.yml) files missing ASF license headers](https://github.com/apache/incubator-openwhisk/issues/3660)
- PR [#3670 - Add ASF Headers to .groovy source files.](https://github.com/apache/incubator-openwhisk/pull/3670)
- PR [#3684 - Update last, edge-case files without ASF headers](https://github.com/apache/incubator-openwhisk/pull/3684)
- PR [#3702 - Add ASF license to optional XML files and last gradle file](https://github.com/apache/incubator-openwhisk/pull/3702)
    -  Reverted by Christian as it broke IBM private PG...
    - See [#3705](https://github.com/apache/incubator-openwhisk/pull/3705) and [#3708](https://github.com/apache/incubator-openwhisk/pull/3708)
- PR [#3703 - Add the license and notice for ConcurrentMapBackedCache.scala](https://github.com/apache/incubator-openwhisk/pull/3703)
    - Issue [#152 - Lightbend copyright appears, no LICENSE attribution](https://github.com/apache/incubator-openwhisk-release/issues/152)
- PR [#3705 - Revert "Add ASF license to optional XML files and last gradle file](https://github.com/apache/incubator-openwhisk/pull/3705)
- PR [#3708 - Single build.gradle file missing ASF license header](https://github.com/apache/incubator-openwhisk/pull/3708)
- PR [#3717 - Update Travis to use ASF-Release.cfg.](https://github.com/apache/incubator-openwhisk/pull/3717)
- PR [#3720 - Update the document regarding Open JDK and Oracle JDK](https://github.com/apache/incubator-openwhisk/pull/3720)

**Related PRs**:
- PR [#155 - Add the tests directory back to the source code release](https://github.com/apache/incubator-openwhisk-release/pull/155), release now includes "tests" sub-directory.
- PR [#154 - Provide per-repo., file-level ASF license header exclusion listing](https://github.com/apache/incubator-openwhisk-release/pull/154)

**Notes**:
Please use this issue to document directories/files (by extension) or individual files we have decided will not require a LICENSE header because it falls under an allowed Apache category (see [https://www.apache.org/legal/src-headers.html](https://www.apache.org/legal/src-headers.html))

**Comment by daisy-ycguo** [comment]
I raised two issues to openwhisk repo to track the license headers https://github.com/apache/incubator-openwhisk/issues/3264 and https://github.com/apache/incubator-openwhisk/issues/3555. Both are in opening status.

**Comment by mrutkows** [comment]
See PR https://github.com/apache/incubator-openwhisk/pull/3550#pullrequestreview-113265457 for the LICENSE file created by @houshengbo

**Comment by mrutkows** [comment]
The question came up about how we should (or even if we should) list services and tooling the project depends upon to build and run.  
For example, we need Ansible and CouchDB (until a Memory Artifactory is completed) (see https://github.com/apache/incubator-openwhisk/blob/master/tools/travis/setup.sh)

**Comment by mrutkows** [comment]
With the noted library use in LICENSE of "logback: which uses LGPL, have we done enough?  After reading the advice in URL below, we address #2, but can we address #1?

https://softwareengineering.stackexchange.com/questions/86142/what-exactly-do-i-need-to-do-if-i-use-a-lgpl-licenced-library/86158

**Comment by daisy-ycguo** [comment]
We should add `*/licenses/*` in the exclusion list.

**Comment by daisy-ycguo** [comment]
Two PRs are created:
https://github.com/apache/incubator-openwhisk/pull/3566
https://github.com/apache/incubator-openwhisk/pull/3565

**Comment by mrutkows** [comment]
@daisy-ycguo @houshengbo Added exclusion for "performance" directory for main openwhisk: 
PR https://github.com/apache/incubator-openwhisk-utilities/pull/26

**Comment by mrutkows** [comment]
@daisy-ycguo Carlos and I were chatting about a PR for scancode and we started talking about the new "performance" directory under main openwhisk.   It now has new dependencies...

here is what we need to look at closely as these may not have public licenses... 

```
Need to see the license on the docker image we use `williamyeh/wrk`
https://hub.docker.com/r/williamyeh/wrk/
```

```
his one doesn’t seem to have a license https://github.com/William-Yeh/docker-wrk/blob/master/Dockerfile
GitHub
William-Yeh/docker-wrk
docker-wrk - A minimal wrk image for Docker - Modern HTTP benchmarking tool.
checking `FROM williamyeh/ansible:mini-alpine3`
```

```
`performance` folder can be deleted and not included in the release .tgz
same for `tests` no need it in the release.tgz
but nice to include if possible
Found it
wrk is Apache 2.0 https://github.com/wg/wrk/blob/master/LICENSE
GitHub
wg/wrk
wrk - Modern HTTP benchmarking tool
```

**Comment by daisy-ycguo** [comment]
@mrutkows I don't quite understand the reason to exclude the whole `performance` folder. Do you want to exclude only `.lua` files from header checking ? or you want to exclude LUA dependency from legal files ?

**Comment by mrutkows** [comment]
@daisy-ycguo what Carlos is saying, and I agree, is that the test tool we are bringing in does not appear to have a license, so we may need to exclude these tests from any release packaging

**Comment by mrutkows** [comment]
(my comment)... perhaps we should move “performance” under “tests” dir. to clarify and simplify our exclusions (both in docs and in release tooling)

**Comment by daisy-ycguo** [comment]
A PR https://github.com/apache/incubator-openwhisk/pull/3596/files is proposed to add short license header to test data files.

I also found that below files should be excluded from license header checking. I will update the document and pom.xml later.
```
  openwhisk/tests/dat/apigw/testswaggerdocinvalid
  openwhisk/tests/dat/apigw/testswaggerdoc2
  openwhisk/tests/dat/apigw/testswaggerdoc1
  openwhisk/tests/src/test/resources/application.conf.j2
```

**Comment by mrutkows** [comment]
@houshengbo is making updates on this PR https://github.com/apache/incubator-openwhisk/pull/3601

**Comment by mrutkows** [comment]
@daisy-ycguo merged PR [Add short license header to test data files](https://github.com/apache/incubator-openwhisk/pull/3596#pullrequestreview-116928826)

**Comment by mrutkows** [comment]
merged PR [#3602 - Update build to use latest scancode .cfg for .md and .sh](https://github.com/apache/incubator-openwhisk/pull/3602#pullrequestreview-116755113)

**Comment by mrutkows** [comment]
merged PR [#3601 - Change the LICENSE file into the correct format defined in openwhisk release](https://github.com/apache/incubator-openwhisk/pull/3601)
